---
## <b> Pandas Challenge - PyCitySchools </b>
#### <i> Submitted by : Sheetal Bongale </i>
##### *This Jupyter notebook will analyze a school district's performance and standardized test results data into meaningful insights.*
---
### Note
* Please find the written analysis of this data at the bottom of this jupyter notebook.


In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas Data Frames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [2]:
# Calculate district's key metrics
total_school_count = school_data_complete["school_name"].nunique()
total_student_count = student_data["student_name"].count()
total_budget = school_data['budget'].sum()

avg_math_score = school_data_complete["math_score"].mean()
avg_reading_score = school_data_complete["reading_score"].mean()

passing_math_percentage = (school_data_complete['math_score'] >= 70).sum()/ total_student_count *100
passing_reading_percentage = (school_data_complete['reading_score'] >= 70).sum()/ total_student_count *100
overall_pass_percentage =  (avg_math_score + avg_reading_score)/2

# Dataframe to store the summary
district_summary = pd.DataFrame({
    "Total Schools": [total_school_count],
    "Total Students": [total_student_count],
    "Total Budget": [total_budget],
    "Average Math Score": [avg_math_score],
    "Average Reading Score": [avg_reading_score],
    "% Passing Math": [passing_math_percentage],
    "% Passing Reading":[passing_reading_percentage],
    "% Overall Passing Rate": [overall_pass_percentage]
})

# Reformat and Output
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)
district_summary


Total Schools  Total Students    Total Budget  Average Math Score  \
0             15           39170  $24,649,428.00           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  \
0               81.87784       74.980853          85.805463   

   % Overall Passing Rate  
0               80.431606

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)
  
* Create a dataframe to hold the above results

In [3]:
# Group by school
groupby_schools = school_data_complete.groupby("school_name")

# Calculations for the School summary
school_type = school_data.set_index(["school_name"])["type"]
total_students = groupby_schools["Student ID"].nunique()
total_school_budget = groupby_schools["budget"].mean()
per_student_budget = total_school_budget/total_students

avg_math_score = groupby_schools["math_score"].mean()
avg_reading_score = groupby_schools["reading_score"].mean()

passing_cutoff_math = school_data_complete[(school_data_complete["math_score"]>=70)]
passing_math_percent = passing_cutoff_math.groupby(["school_name"]).count()["student_name"]/total_students *100
passing_cutoff_read = school_data_complete[(school_data_complete["reading_score"]>=70)]
passing_reading_percent = passing_cutoff_read.groupby(["school_name"]).count()["student_name"]/total_students *100

overall_passing_rate = (passing_math_percent + passing_reading_percent)/2

# Dataframe to store the summary
school_summary = pd.DataFrame({
    "School Type": school_type,
    "Total Students": total_students,
    "Total School Budget": total_school_budget,
    "Per Student Budget" : per_student_budget,
    "Average Math Score": avg_math_score,
    "Average Reading Score": avg_reading_score,
    "% Passing Math": passing_math_percent,
    "% Passing Reading":passing_reading_percent,
    "% Overall Passing Rate": overall_passing_rate
})

# Reformat and Output
formatted_school_summary = school_summary.copy()
columns = ["Total School Budget","Per Student Budget"]
formatted_school_summary[columns] = formatted_school_summary[columns].applymap("${:,.2f}".format)
formatted_school_summary


School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
Bailey High School               $628.00           77.048432   
Cabrera High School              $582.00           83.061895   
Figueroa High School             $639.00           76.711767   
Ford High School                 $644.00           77.102592   
Griffin High School              $625.00           83.351499   
Hernandez High School            $652.00           77.289752   
Holden High School               $581.00           83.803279   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Pena High School                 $609.00           83.839917   
Rodriguez High School            $637.00           76.842711   
Shelton High School              $600.00           83.359455   
Thomas High School               $638.00           83.418349   
Wilson High School               $578.00           83.274201   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing Rate  
Bailey High School             81.933280               74.306672  
Cabrera High School            97.039828               95.586652  
Figueroa High School           80.739234               73.363852  
Ford High School               79.299014               73.804308  
Griffin High School            97.138965               95.265668  
Hernandez High School          80.862999               73.807983  
Holden High School             96.252927               94.379391  
Huang High School              81.316421               73.500171  
Johnson High School            81.222432               73.639992  
Pena High School               95.945946               95.270270  
Rodriguez High School          80.220055               73.293323  

## Top Performing Schools (By Passing Rate)

* Sort and display the top five schools in overall passing rate

In [4]:
# Top five schools by overall passing rate
formatted_school_summary.sort_values("% Overall Passing Rate", ascending=False).head(5)


School Type  Total Students Total School Budget  \
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Pena High School        Charter             962         $585,858.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   

                    Per Student Budget  Average Math Score  \
Cabrera High School            $582.00           83.061895   
Thomas High School             $638.00           83.418349   
Pena High School               $609.00           83.839917   
Griffin High School            $625.00           83.351499   
Wilson High School             $578.00           83.274201   

                     Average Reading Score  % Passing Math  % Passing Reading  \
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Pena High School                 84.044699       94.594595          95.945946   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   

                     % Overall Passing Rate  
Cabrera High School               95.586652  
Thomas High School                95.290520  
Pena High School                  95.270270  
Griffin High School               95.265668  
Wilson High School                95.203679

## Bottom Performing Schools (By Passing Rate)

* Sort and display the five worst-performing schools

In [5]:
# Worst five schools by overall passing rate
formatted_school_summary.sort_values("% Overall Passing Rate", ascending=True).head(5)


School Type  Total Students Total School Budget  \
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Ford High School         District            2739       $1,763,916.00   

                      Per Student Budget  Average Math Score  \
Rodriguez High School            $637.00           76.842711   
Figueroa High School             $639.00           76.711767   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Ford High School                 $644.00           77.102592   

                       Average Reading Score  % Passing Math  \
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   

                       % Passing Reading  % Overall Passing Rate  
Rodriguez High School          80.220055               73.293323  
Figueroa High School           80.739234               73.363852  
Huang High School              81.316421               73.500171  
Johnson High School            81.222432               73.639992  
Ford High School               79.299014               73.804308

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [6]:
# Math scores for students of each grade level at each school
grade_and_school = school_data_complete.groupby(["school_name", "grade"])
math_scores = grade_and_school["math_score"].mean().unstack()
math_scores = math_scores[["9th", "10th", "11th", "12th"]]

# Format the column index name and Output
math_scores.columns.name = math_scores.index.name = None
math_scores

9th       10th       11th       12th
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [7]:
# Reading scores for students of each grade level at each school
reading_scores = grade_and_school["math_score"].mean().unstack()
reading_scores = reading_scores[["9th", "10th", "11th", "12th"]]

# Format the column index name and Output
reading_scores.columns.name = reading_scores.index.name = None
reading_scores

9th       10th       11th       12th
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [13]:
# Sample bins. Feel free to create your own bins.
spending_bins = [0, 585, 615, 645, 675]
group_names = ["<$585", "$585-615", "$615-645", "$645-675"]

In [14]:
# Set columns and add to bins
scores_by_school_spending = school_summary.loc[:,["Average Math Score",
                                                  "Average Reading Score",
                                                  "% Passing Math",
                                                  "% Passing Reading",
                                                  "% Overall Passing Rate"]]

scores_by_school_spending["Spending Ranges (Per Student)"] = pd.cut(
    school_summary["Per Student Budget"], spending_bins, labels=group_names
)
scores_by_school_spending = scores_by_school_spending.groupby("Spending Ranges (Per Student)").mean()
scores_by_school_spending

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                   83.455399              83.933814   
$585-615                                83.599686              83.885211   
$615-645                                79.079225              81.891436   
$645-675                                76.997210              81.027843   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                               93.460096          96.610877   
$585-615                            94.230858          95.900287   
$615-645                            75.668212          86.106569   
$645-675                            66.164813          81.133951   

                               % Overall Passing Rate  
Spending Ranges (Per Student)                          
<$585                                       95.035486  
$585-615                                    95.065572  
$615-645                                    80.887391  
$645-675                                    73.649382

## Scores by School Size

* Perform the same operations as above, based on school size.

In [10]:
# Sample bins. Feel free to create your own bins.
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

scores_by_school_size = school_summary.loc[:,["Average Math Score",
                                                  "Average Reading Score",
                                                  "% Passing Math",
                                                  "% Passing Reading",
                                                  "% Overall Passing Rate"]]

scores_by_school_size["School Size"] = pd.cut(
    school_summary["Total Students"], size_bins, labels=group_names
)
scores_by_school_size = scores_by_school_size.groupby("School Size").mean()
scores_by_school_size

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing Rate  
School Size                                                    
Small (<1000)               96.099437               94.824831  
Medium (1000-2000)          96.790680               95.195187  
Large (2000-5000)           82.766634               76.364998

## Scores by School Type

* Perform the same operations as above, based on school type.

In [11]:
scores_by_school_type = school_summary.loc[:,["School Type",
                                        "Average Math Score",
                                        "Average Reading Score",
                                        "% Passing Math",
                                        "% Passing Reading",
                                        "% Overall Passing Rate"]]

scores_by_school_type = scores_by_school_type.groupby('School Type').mean()
scores_by_school_type

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing Rate  
School Type                                             
Charter              96.586489               95.103660  
District             80.799062               73.673757

---
# School District Data Analysis and Insights:
---
Following were the trends observed after the analysis:
- **Average Scores**: In all 15 schools, the ***Average Reading Score*** is higher than the **Average Math Score.** Average math and reading scores stayed
    consistent across grade level.

- **School Size**: ***Small*** and ***Medium schools*** had a overall passing rate of ***94.82%*** and ***95.19%*** respectively compared to ***76.36%*** for
    ***Large schools***, showing us a direct correlation between the ***school size and school performance***. 
    Also, small and medium sized schools dramatically out-performed large sized schools in ***math test performance (93.5% vs 69%)***. 

- **School Type**: The top five performing schools were all ***Charter Schools*** with an overall passing rate of ***95.10%***. Also the correlation between the
    school size and the school performance is valid here because the charter schools have sigficantly ***lower student counts***.

- **School Budget**: ***District schools*** seemed to have ***higher budgets per student*** than charter schools; but did not yield better test results. Hence, there is not neccessarily a correlation between school spending and test performance.
---